In this exercise, we will first create an extension of the IoU- and ReID-based tracker we created in the previous exercise that will make it more robust to occlusions by allowing it to recover from missed detections.

We will then implement a Message Passing Network that we will use to learn features that combine position and appearance information. On these edge features, we then base our prediction on associations between past tracks and new detections.

Your tasks are the following:
- Adapt the track management scheme of our Tracker to allow it to recover from missed detections.
- Implement a Message Passing Network to operate on bipartite graphs
- Implement the pairwise feature computation to obtain features for our Message Passing Network
- Train the Message Passing Network and improve your tracker's IDF1 score
- Use the Message Passing Network instead of the distance matrix in the Tracker

##  Setup
### Download and extract project data to your Google Drive

1. **Required**: Please follow all instructions of exercise 0 and exercise 1 before running this notebook.
1. Save this notebook to your Google Drive by clicking Save a copy in Drive from the File menu.
3. Download [this](https://vision.in.tum.de/webshare/u/cetintas/cv3dst_gnn_exercise.zip) zip file to your desktop, extract it and upload it into the Colab Notebooks folder in your Google Drive.

#### Connect the notebook to your Google Drive

In [1]:
#from google.colab import drive

#drive.mount('/content/gdrive')

In [1]:
root_dir = "cv3dst_exercise/"
gnn_root_dir = "cv3dst_gnn_exercise/"

The `root_dir` path points to the directory and the content in your Google Drive.

In [2]:
!ls "gdrive/My Drive/Colab Notebooks/cv3dst_gnn_exercise/data"

'ls' is not recognized as an internal or external command,
operable program or batch file.


#### Install and import Python libraries

In [4]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

!pip install tqdm lap
!pip install https://github.com/timmeinhardt/py-motmetrics/archive/fix_pandas_deprecating_warnings.zip

  Using cached https://github.com/timmeinhardt/py-motmetrics/archive/fix_pandas_deprecating_warnings.zip
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'


In [5]:
import os
import sys
sys.path.append(os.path.join(gnn_root_dir, 'src'))


import matplotlib.pyplot as plt
import numpy as np
import time
from tqdm.autonotebook import tqdm

import torch
from torch.utils.data import DataLoader

from tracker.data_track import MOT16Sequences
from tracker.tracker import Tracker, ReIDTracker
from tracker.utils import run_tracker, cosine_distance
from scipy.optimize import linear_sum_assignment as linear_assignment
import os.path as osp

import motmetrics as mm
mm.lap.default_solver = 'lap'

C:\Users\yusuf\AppData\Local\Temp\ipykernel_26116\3896152220.py:9: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm


In [6]:
!ls "gdrive/My Drive/Colab Notebooks/cv3dst_exercise/data/MOT16/train"
!ls "gdrive/My Drive/Colab Notebooks/cv3dst_exercise/data/MOT16/test"

'ls' is not recognized as an internal or external command,
operable program or batch file.
'ls' is not recognized as an internal or external command,
operable program or batch file.


## Speed-Ups
In order to speed up training and inference runtimes, in this exercise we will be working with pre-computed detections and ReID embeddings. We ran the object detector we provided in Exercise 0 and applied to all frames. We also computed reid embeddings for all boxes in every frame of the dataset so that they don't need to be computed every time you run your tracker. This yields over 10x speed improvements. You will not have to work directly with the resulting files, as we have internally adapted the boilerplate code to work with them.

In [7]:
train_db = torch.load(osp.join(gnn_root_dir, 'data/preprocessed_data_train_2.pth'))

## Exercise Part 0 - Assignment's 1 ReIDHungarianTracker

We start by providing a sample solution of the ``ReIDTracker`` from Exercise 1.
It will serve as our baseline.


Recall that this tracker works by performing frame-to-frame bipartite matching between newly detected boxes and past tracks based on ReID distance. Whenever a past track cannot be matched, its killed. And whenever, a newly detected box cannot be match, it starts a new trajectory.

**NOTE**: We have modified the ``compute_distance`` function in ``data_association`` from last week to include a threshold on ReID distance (if ReID distance >0.1, matching is not possible). This is important to prevent our tracker from reusing tracks for very dissimilar objects.

Results:
```
          IDF1   IDP   IDR  Rcll  Prcn  GT  MT  PT ML   FP    FN IDs   FM  MOTA  MOTP
MOT16-02 40.5% 57.5% 31.3% 52.2% 96.1%  62  12  38 12  390  8873 203  210 49.1% 0.090
MOT16-05 54.4% 64.4% 47.1% 68.8% 94.0% 133  54  67 12  305  2156 330  149 59.7% 0.142
MOT16-09 49.9% 61.7% 41.9% 66.3% 97.7%  26  13  12  1   82  1793  77   76 63.3% 0.082
MOT16-11 61.1% 67.4% 55.9% 80.2% 96.6%  75  44  24  7  266  1871 176   91 75.5% 0.083
OVERALL  49.6% 62.3% 41.2% 63.5% 96.1% 296 123 141 32 1043 14693 786  526 59.0% 0.097
```


In [8]:
_UNMATCHED_COST=255
class ReIDHungarianTracker(ReIDTracker):
    def data_association(self, boxes, scores, pred_features):
        """Refactored from previous implementation to split it onto distance computation and track management"""
        if self.tracks:
            track_boxes = torch.stack([t.box for t in self.tracks], axis=0)
            track_features = torch.stack([t.get_feature() for t in self.tracks], axis=0)

            distance = self.compute_distance_matrix(track_features, pred_features,
                                                    track_boxes, boxes, metric_fn=cosine_distance)

            # Perform Hungarian matching.
            row_idx, col_idx = linear_assignment(distance)
            self.update_tracks(row_idx, col_idx, distance, boxes, scores, pred_features)


        else:
            # No tracks exist.
            self.add(boxes, scores, pred_features)

    def update_tracks(self, row_idx, col_idx, distance, boxes, scores, pred_features):
        """Updates existing tracks and removes unmatched tracks.
           Reminder: If the costs are equal to _UNMATCHED_COST, it's not a
           match.
        """
        track_ids = [t.id for t in self.tracks]

        unmatched_track_ids = []
        seen_track_ids = []
        seen_box_idx = []
        for track_idx, box_idx in zip(row_idx, col_idx):
            costs = distance[track_idx, box_idx]
            internal_track_id = track_ids[track_idx]
            seen_track_ids.append(internal_track_id)
            if costs == _UNMATCHED_COST:
                unmatched_track_ids.append(internal_track_id)
            else:
                self.tracks[track_idx].box = boxes[box_idx]
                self.tracks[track_idx].add_feature(pred_features[box_idx])
                seen_box_idx.append(box_idx)

        unseen_track_ids = set(track_ids) - set(seen_track_ids)
        unmatched_track_ids.extend(list(unseen_track_ids))
        self.tracks = [t for t in self.tracks
                       if t.id not in unmatched_track_ids]


        # Add new tracks.
        new_boxes_idx = set(range(len(boxes))) - set(seen_box_idx)
        new_boxes = [boxes[i] for i in new_boxes_idx]
        new_scores = [scores[i] for i in new_boxes_idx]
        new_features = [pred_features[i] for i in new_boxes_idx]
        self.add(new_boxes, new_scores, new_features)


In [9]:
val_sequences = MOT16Sequences('MOT16-reid', root_dir = osp.join(root_dir, 'data/MOT16'), vis_threshold=0.)

In [10]:
tracker = ReIDHungarianTracker(None)
run_tracker(val_sequences, db=train_db, tracker=tracker, output_dir=None)

Tracking: MOT16-02
Tracks found: 367
Runtime for MOT16-02: 6.4 s.
Tracking: MOT16-05
Tracks found: 464
Runtime for MOT16-05: 1.8 s.
Tracking: MOT16-09
Tracks found: 98
Runtime for MOT16-09: 1.1 s.
Tracking: MOT16-11
Tracks found: 292
Runtime for MOT16-11: 2.5 s.
Runtime for all sequences: 11.7 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT  MT  PT ML   FP    FN IDs   FM  MOTA  MOTP
MOT16-02 40.5% 57.5% 31.3% 52.2% 96.1%  62  12  38 12  390  8873 203  210 49.1% 0.090
MOT16-05 54.4% 64.4% 47.1% 68.8% 94.0% 133  54  67 12  305  2156 330  149 59.7% 0.142
MOT16-09 49.9% 61.7% 41.9% 66.3% 97.7%  26  13  12  1   82  1793  77   76 63.3% 0.082
MOT16-11 61.1% 67.4% 55.9% 80.2% 96.6%  75  44  24  7  266  1871 176   91 75.5% 0.083
OVERALL  49.6% 62.3% 41.2% 63.5% 96.1% 296 123 141 32 1043 14693 786  526 59.0% 0.097


idf1       idp       idr    recall  precision  \
MOT16-02  0.405105  0.575262  0.312631  0.522469   0.961378   
MOT16-05  0.544104  0.643506  0.471303  0.688304   0.939795   
MOT16-09  0.499161  0.617322  0.418967  0.663286   0.977310   
MOT16-11  0.611340  0.673988  0.559347  0.801717   0.966032   
OVERALL   0.495843  0.623022  0.411784  0.635038   0.960803   

          num_unique_objects  mostly_tracked  partially_tracked  mostly_lost  \
MOT16-02                  62              12                 38           12   
MOT16-05                 133              54                 67           12   
MOT16-09                  26              13                 12            1   
MOT16-11                  75              44                 24            7   
OVERALL                  296             123                141           32   

          num_false_positives  num_misses  num_switches  num_fragmentations  \
MOT16-02                  390        8873           203                 210   
MOT16-05                  305        2156           330                 149   
MOT16-09                   82        1793            77                  76   
MOT16-11                  266        1871           176                  91   
OVERALL                  1043       14693           786                 526   

              mota      motp  
MOT16-02  0.490555  0.090340  
MOT16-05  0.596501  0.142002  
MOT16-09  0.633427  0.082336  
MOT16-11  0.754875  0.082736  
OVERALL   0.589607  0.096605

## Exercise Part I - Long-Term ReID Tracker


The tracker above has an obvious limitation: whenever a track cannot be matched with the detections of a given frame the track will be killed. This means that if our detector misses an object in a single frame (due to e.g. occlusion), we will not be able to recover that track, and we will start a new one.

To fix this issue, we would like to allow our tracker to maintain tracks that are not matched during data association. We will refer to these tracks as **inactive**. During data association, we will try to match the detected boxes for the current frame to both tracks that are active (i.e. tracks that we were able to match in the previous frame) as well as those that are inactive. Therefore, if a detector misses an object in a frame and the object reappears after a few frames, we will still be able to match it to its corresponding track, instead of creating a new one.

In order to adapt our tracker to have this behavior, we will use the `inactive` attribute from the `track` class (see `tracker/tracker.py`. This attribute will be assigned an integer indicating for how many frames a track has remained unmatched. Whenever we are able to match the track `t`, we will set `t.inactive=0` and, naturally, when tracks are initialized, the class constructor sets `inactive=0`.

Your job is to maintain the `inactive` attribute of all tracks being kept by tracker so that its value represents the number of frames for which the track has been unmatched. Additionally, we introduce a `patience` parameter. Whenever a track has been inactive for more than `inactive` frames. it will need to be killed.

Results should approximately be around:

```
          IDF1   IDP   IDR  Rcll  Prcn  GT  MT  PT ML   FP    FN IDs   FM  MOTA  MOTP
MOT16-02 45.9% 65.1% 35.4% 52.2% 96.1%  62  12  37 13  390  8873 130  210 49.4% 0.090
MOT16-05 63.4% 75.0% 54.9% 68.8% 94.0% 133  54  67 12  305  2156 283  149 60.3% 0.142
MOT16-09 52.5% 64.9% 44.1% 66.3% 97.7%  26  13  12  1   82  1793  49   76 63.9% 0.083
MOT16-11 68.3% 75.3% 62.5% 80.2% 96.6%  75  44  24  7  266  1871 136   90 75.9% 0.083
OVERALL  55.7% 70.0% 46.2% 63.5% 96.1% 296 123 140 33 1043 14693 598  525 59.4% 0.097
```


In [11]:
class LongTermReIDHungarianTracker(ReIDHungarianTracker):
    def __init__(self, patience, *args, **kwargs):
        """ Add a patience parameter"""
        self.patience=patience
        super().__init__(*args, **kwargs)

    def update_results(self):
        """Only store boxes for tracks that are active"""
        for t in self.tracks:
            if t.id not in self.results.keys():
                self.results[t.id] = {}
            if t.inactive == 0: # Only change
                self.results[t.id][self.im_index] = np.concatenate([t.box.cpu().numpy(), np.array([t.score])])

        self.im_index += 1

    def update_tracks(self, row_idx, col_idx, distance, boxes, scores, pred_features):
        track_ids = [t.id for t in self.tracks]

        unmatched_track_ids = []
        seen_track_ids = []
        seen_box_idx = []
        for track_idx, box_idx in zip(row_idx, col_idx):
            costs = distance[track_idx, box_idx]
            internal_track_id = track_ids[track_idx]
            seen_track_ids.append(internal_track_id)
            if costs == _UNMATCHED_COST:
                unmatched_track_ids.append(internal_track_id)

            else:
                self.tracks[track_idx].box = boxes[box_idx]
                self.tracks[track_idx].add_feature(pred_features[box_idx])

                # Note: the track is matched, therefore, inactive is set to 0
                self.tracks[track_idx].inactive=0
                seen_box_idx.append(box_idx)


        unseen_track_ids = set(track_ids) - set(seen_track_ids)
        unmatched_track_ids.extend(list(unseen_track_ids))
        ##################
        ### TODO starts
        ##################

        # Update the `inactive` attribute for those tracks that have been
        # not been matched. kill those for which the inactive parameter
        # is > self.patience
        for track in self.tracks:
            if track.id in unseen_track_ids:
                track.inactive += 1
                if track.inactive > self.patience:
                    self.tracks.remove(track)



        ##################
        ### TODO ends
        ##################

        new_boxes_idx = set(range(len(boxes))) - set(seen_box_idx)
        new_boxes = [boxes[i] for i in new_boxes_idx]
        new_scores = [scores[i] for i in new_boxes_idx]
        new_features = [pred_features[i] for i in new_boxes_idx]
        self.add(new_boxes, new_scores, new_features)

In [12]:
tracker = LongTermReIDHungarianTracker(patience=20, obj_detect=None)
run_tracker(val_sequences, db=train_db, tracker=tracker, output_dir=None)

Tracking: MOT16-02
Tracks found: 207
Runtime for MOT16-02: 7.4 s.
Tracking: MOT16-05
Tracks found: 354
Runtime for MOT16-05: 2.1 s.
Tracking: MOT16-09
Tracks found: 64
Runtime for MOT16-09: 1.3 s.
Tracking: MOT16-11
Tracks found: 217
Runtime for MOT16-11: 3.0 s.
Runtime for all sequences: 13.8 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT  MT  PT ML   FP    FN IDs   FM  MOTA  MOTP
MOT16-02 45.9% 65.1% 35.5% 52.3% 96.1%  62  12  37 13  397  8857 130  208 49.5% 0.091
MOT16-05 63.3% 74.6% 55.1% 68.9% 93.3% 133  54  67 12  341  2149 283  149 59.9% 0.142
MOT16-09 52.6% 65.0% 44.2% 66.4% 97.6%  26  12  13  1   86  1789  48   78 63.9% 0.087
MOT16-11 68.2% 74.9% 62.5% 80.2% 96.1%  75  44  24  7  305  1866 136   89 75.6% 0.083
OVERALL  55.7% 69.8% 46.3% 63.6% 95.8% 296 122 141 33 1129 14661 597  524 59.3% 0.097


idf1       idp       idr    recall  precision  \
MOT16-02  0.458992  0.650825  0.354502  0.523330   0.960775   
MOT16-05  0.633461  0.745547  0.550672  0.689316   0.933255   
MOT16-09  0.526433  0.650193  0.442254  0.664038   0.976256   
MOT16-11  0.681879  0.749460  0.625477  0.802247   0.961270   
OVERALL   0.556922  0.697908  0.463325  0.635833   0.957758   

          num_unique_objects  mostly_tracked  partially_tracked  mostly_lost  \
MOT16-02                  62              12                 37           13   
MOT16-05                 133              54                 67           12   
MOT16-09                  26              12                 13            1   
MOT16-11                  75              44                 24            7   
OVERALL                  296             122                141           33   

          num_false_positives  num_misses  num_switches  num_fragmentations  \
MOT16-02                  397        8857           130                 208   
MOT16-05                  341        2149           283                 149   
MOT16-09                   86        1789            48                  78   
MOT16-11                  305        1866           136                  89   
OVERALL                  1129       14661           597                 524   

              mota      motp  
MOT16-02  0.494968  0.090509  
MOT16-05  0.599104  0.142270  
MOT16-09  0.638873  0.087220  
MOT16-11  0.755511  0.082846  
OVERALL   0.592961  0.097430

## Exercise Part II - Building a tracker based on Neural Message Passing

Our ``LongTermReIDHungarianTracker`` is still limited when compared to current modern trackers.

Firstly, it relies solely on appearance to predict similarity scores between objectes. This can be problematic whenever appearance alone may not discriminative, and it'd be best to also take into account object position and size attributes. Secondly, our tracker can only account for pairwise similarities among objects. Ideally, we would like it to also consider higher-order information.

To address these limitations. We will now build a tracker that will combine both apperance and position information with a Message Passing Neural Network, inspired by the approach presented in [Learning a Neural Solver for Multiple Object Tracking, CVPR 2020](https://arxiv.org/abs/1912.07515)

The overall idea will be to build, for every tracking step, a bipartite graph containing two sets of nodes: past tracks, and detections in the current frame. We will initialize node features with ReID embeddings, and edge features with relative position features and ReID distance. We will use an MPN to refine these edge embeddings. The learning task will be to classify the edge embeddings in this graph, which is equivalent to predicting the entries of our data association similarity matrix.


### Building an MPN for Bipartite Graphs

We will first build a Neural Message Passing layer based on the Graph Networks framework introduced in [Relational inductive biases, deep learning, and graph networks, arXiv 2020](https://arxiv.org/abs/1806.01261), as explained in the *A More General Framework* slides of [Lecture 5](https://youtu.be/BR3Y5bAz5Dw) (slides 70 to 75).

We will be using a bipartite graph, i.e., we will have two sets of nodes $A$ (past tracks), and $B$ (detections), and our set of edges will be $A\times B$. That is, we will connect every pair of past tracks and detections.

We will have initial node features (i.e. reid embeddings) matrices: $X_A$ and $X_B$ and an initial edge features tensor $E$.

$X_A$ and $X_B$ have shape $|A|\times \text{node_dim}$ and $|B|\times \text{node_dim}$, respectively.

$E$ has shape $|A| \times |B| \times \text{edge_dim}$. Its $(i, j)$ entry contains the edge features of node $i$ in $A$ and node $j$ in $B$.

With the given layer, we will produce new node feature matrices $X_A'$ and $X_B'$ and edge features $E'$ with the same dimensions.
Please refer to the formulas in the slides and figure how to apply them in this setting.

You are asked to implement both the node and edge update steps in the class below

**NOTE 1**: Working with a bipartite graph allows us to vectorize all operations in the formulas in a straightforward manner (keep in mind that we store edge features in a matrix). Given a node in $A$, it is connected to all nodes in $B$.

**NOTE 2**: You do not need to care about batching several graphs. This implementation will only work with a single graph at a time.

In [13]:
from torch import nn

class BipartiteNeuralMessagePassingLayer(nn.Module):
    def __init__(self, node_dim, edge_dim, dropout=0.):
        super().__init__()

        edge_in_dim  = 2*node_dim + 2*edge_dim # 2*edge_dim since we always concatenate initial edge features
        self.edge_mlp = nn.Sequential(*[nn.Linear(edge_in_dim, edge_dim), nn.ReLU(), nn.Dropout(dropout),
                                    nn.Linear(edge_dim, edge_dim), nn.ReLU(), nn.Dropout(dropout)])

        node_in_dim  = node_dim + edge_dim
        self.node_mlp = nn.Sequential(*[nn.Linear(node_in_dim, node_dim), nn.ReLU(), nn.Dropout(dropout),
                                        nn.Linear(node_dim, node_dim), nn.ReLU(), nn.Dropout(dropout)])

    def edge_update(self, edge_embeds, nodes_a_embeds, nodes_b_embeds):
        """
        Node-to-edge updates, as descibed in slide 71, lecture 5.
        Args:
            edge_embeds: torch.Tensor with shape (|A|, |B|, 2 x edge_dim)
            nodes_a_embeds: torch.Tensor with shape (|A|, node_dim)
            nodes_a_embeds: torch.Tensor with shape (|B|, node_dim)

        returns:
            updated_edge_feats = torch.Tensor with shape (|A|, |B|, edge_dim)
        """

        n_nodes_a, n_nodes_b, _  = edge_embeds.shape

        ########################
        #### TODO starts
        ########################
        _, node_dim = nodes_a_embeds.shape
        #nodes_a_embeds = nodes_a_embeds.repeat(n_nodes_b,1).reshape(n_nodes_a,n_nodes_b,node_dim ) # a ,b , node_dim
        #nodes_b_embeds = nodes_b_embeds.repeat(n_nodes_a,1).reshape(n_nodes_a,n_nodes_b,node_dim ) # a ,b, node_dim
        #edge_in = torch.cat( (nodes_a_embeds, nodes_b_embeds, edge_embeds) , axis=2) # edge a,b, 2*c

        nodes_a_embeds_repeated =  nodes_a_embeds.reshape((n_nodes_a, 1, node_dim)).repeat_interleave(n_nodes_b, axis = 1)  # |A|, |B|, node_dim
        nodes_b_embeds_repeated =  nodes_b_embeds.reshape((1, n_nodes_b, node_dim)).repeat_interleave(n_nodes_a, axis = 0)  # |A|, |B|, node_dim
        edge_in = torch.cat((nodes_a_embeds_repeated, edge_embeds, nodes_b_embeds_repeated), 2) # |A|, |B|, 2*edge_dim + 2*node_dim
        ########################
        #### TODO ends
        ########################


        return self.edge_mlp(edge_in)

    def node_update(self, edge_embeds, nodes_a_embeds, nodes_b_embeds):
        """
        Edge-to-node updates, as descibed in slide 75, lecture 5.

        Args:
            edge_embeds: torch.Tensor with shape (|A|, |B|, edge_dim)
            nodes_a_embeds: torch.Tensor with shape (|A|, node_dim)
            nodes_b_embeds: torch.Tensor with shape (|B|, node_dim)

        returns:
            tuple(
                updated_nodes_a_embeds: torch.Tensor with shape (|A|, node_dim),
                updated_nodes_b_embeds: torch.Tensor with shape (|B|, node_dim)
                )
        """

        ########################
        #### TODO starts
        ########################

        # NOTE: Use 'sum' as aggregation function

        nodes_a, nodes_b, edge_dim = edge_embeds.shape
        a_edges = torch.sum(edge_embeds, 1).reshape((nodes_a, edge_dim)) # b , edge
        b_edges = torch.sum(edge_embeds, 0).reshape((nodes_b, edge_dim)) # a , edge
        nodes_a_in = torch.cat((nodes_a_embeds, a_edges), 1)  #  a, edge + node
        nodes_b_in = torch.cat((nodes_b_embeds, b_edges), 1) # b, edge + node

        ########################
        #### TODO ends
        ########################

        nodes_a = self.node_mlp(nodes_a_in)
        nodes_b = self.node_mlp(nodes_b_in)

        return nodes_a, nodes_b

    def forward(self, edge_embeds, nodes_a_embeds, nodes_b_embeds):
        edge_embeds_latent = self.edge_update(edge_embeds, nodes_a_embeds, nodes_b_embeds)
        nodes_a_latent, nodes_b_latent = self.node_update(edge_embeds_latent, nodes_a_embeds, nodes_b_embeds)

        return edge_embeds_latent, nodes_a_latent, nodes_b_latent

## Building the entire network to predict similarities
We now build the network that generates initial node and edge features, performs neural message passing, and classifies edges in order to produce the final costs that we will use for data association.

You need to implement the method that computes the initial edge features. You can follow [1] and, given a two bounding boxes $(x_i, y_i, w_i, h_i)$ and  $(x_j, y_j, w_j, h_j)$ and timestamps $t_i$ and $t_j$, compute an initial 5-dimensional edge feature vector as:
$$ E_{(i, j)} = \left(\frac{2(x_j - x_i)}{h_i + h_j}, \frac{2(y_j - y_i)}{h_i + h_j}, \log{\frac{h_i}{h_j}}, \log{\frac{w_i}{w_j}}, t_j - t_i \right )$$


Feel free to engineer your own features (e.g. use IoU, etc.)

In [14]:
from torch.nn import functional as F
class AssignmentSimilarityNet(nn.Module):
    def __init__(self, reid_network, node_dim, edge_dim, reid_dim, edges_in_dim, num_steps, dropout=0.):
        super().__init__()
        self.reid_network = reid_network
        self.graph_net = BipartiteNeuralMessagePassingLayer(node_dim=node_dim, edge_dim=edge_dim, dropout=dropout)
        self.num_steps = num_steps
        self.cnn_linear = nn.Linear(reid_dim, node_dim)
        self.edge_in_mlp = nn.Sequential(*[nn.Linear(edges_in_dim, edge_dim), nn.ReLU(), nn.Dropout(dropout), nn.Linear(edge_dim, edge_dim), nn.ReLU(),nn.Dropout(dropout)])
        self.classifier = nn.Sequential(*[nn.Linear(edge_dim, edge_dim), nn.ReLU(), nn.Linear(edge_dim, 1)])


    def compute_edge_feats(self, track_coords, current_coords, track_t, curr_t):
        """
        Computes initial edge feature tensor

        Args:
            track_coords: track's frame box coordinates, given by top-left and bottom-right coordinates
                          torch.Tensor with shape (num_tracks, 4)
            current_coords: current frame box coordinates, given by top-left and bottom-right coordinates
                            has shape (num_boxes, 4)

            track_t: track's timestamps, torch.Tensor with with shape (num_tracks, )
            curr_t: current frame's timestamps, torch.Tensor withwith shape (num_boxes,)


        Returns:
            tensor with shape (num_trakcs, num_boxes, 5) containing pairwise
            position and time difference features
        """

        ########################
        #### TODO starts
        ########################

        # NOTE 1: we recommend you to use box centers to compute distances
        # in the x and y coordinates.

        # NOTE 2: Check out the  code inside train_one_epoch function and
        # LongTrackTrainingDataset class a few cells below to debug this
        def get_iou(ground_truth, pred):
            # coordinates of the area of intersection.
            ix1 = np.maximum(ground_truth[0], pred[0])
            iy1 = np.maximum(ground_truth[1], pred[1])
            ix2 = np.minimum(ground_truth[2], pred[2])
            iy2 = np.minimum(ground_truth[3], pred[3])

            # Intersection height and width.
            i_height = np.maximum(iy2 - iy1 + 1, np.array(0.))
            i_width = np.maximum(ix2 - ix1 + 1, np.array(0.))

            area_of_intersection = i_height * i_width

            # Ground Truth dimensions.
            gt_height = ground_truth[3] - ground_truth[1] + 1
            gt_width = ground_truth[2] - ground_truth[0] + 1

            # Prediction dimensions.
            pd_height = pred[3] - pred[1] + 1
            pd_width = pred[2] - pred[0] + 1

            area_of_union = gt_height * gt_width + pd_height * pd_width - area_of_intersection

            iou = area_of_intersection / area_of_union

            return iou
        n_a, _ = track_coords.shape
        n_b, _ = current_coords.shape
        edge_feats = torch.zeros((n_a, n_b, 6))
        track_coords = track_coords.reshape((n_a, 1, 4)).repeat_interleave(n_b, axis=1)  # Making it (n_a, n_b, 4)
        current_coords = current_coords.reshape((1, n_b, 4)).repeat_interleave(n_a, axis=0)  # Making it (n_a, n_b, 4)
        track_t = track_t.reshape((n_a, 1, 1)).repeat_interleave(n_b, axis=1)  # Making it (n_a, n_b, 1)
        curr_t = curr_t.reshape((1, n_b, 1)).repeat_interleave(n_a, axis=0)  # Making it (n_a, n_b, 1)

        def get_center(x1, x2):
            return (x1 + x2) / 2

        c_x = get_center(current_coords[:, :, 0], current_coords[:, :, 2])  # Mean X
        c_y = get_center(current_coords[:, :, 1], current_coords[:, :, 3])  # Mean Y
        c_width = current_coords[:, :, 2] - current_coords[:, :, 0]  # width
        c_height = current_coords[:, :, 3] - current_coords[:, :, 1]  # height

        t_x = get_center(track_coords[:, :, 0], track_coords[:, :, 2])  # Mean X
        t_y = get_center(track_coords[:, :, 1], track_coords[:, :, 3])  # Mean Y
        t_width = track_coords[:, :, 2] - track_coords[:, :, 0]  # width
        t_height = track_coords[:, :, 3] - track_coords[:, :, 1]  # height

        edge_feats[:, :, 0] = 2 * (c_x - t_x) / (c_height + t_height)
        edge_feats[:, :, 1] = 2 * (c_y - t_y) / (c_height + t_height)
        edge_feats[:, :, 2] = torch.log(c_height / t_height)
        edge_feats[:, :, 3] = torch.log(c_width / t_width)
        edge_feats[:, :, 4] = curr_t[:, :, 0] - track_t[:, :, 0]
        edge_feats[:, :, 5] = torch.tensor(
            [get_iou(c, t) for c, t in zip(current_coords.reshape(-1, 4).to('cpu'), track_coords.reshape(-1, 4).to('cpu'))]).reshape(n_a, n_b)




        ########################
        #### TODO ends
        ########################

        return edge_feats # has shape (num_trakcs, num_boxes, 5)


    def forward(self, track_app, current_app, track_coords, current_coords, track_t, curr_t):
        """
        Args:
            track_app: track's reid embeddings, torch.Tensor with shape (num_tracks, 512)
            current_app: current frame detections' reid embeddings, torch.Tensor with shape (num_boxes, 512)
            track_coords: track's frame box coordinates, given by top-left and bottom-right coordinates
                          torch.Tensor with shape (num_tracks, 4)
            current_coords: current frame box coordinates, given by top-left and bottom-right coordinates
                            has shape (num_boxes, 4)

            track_t: track's timestamps, torch.Tensor with with shape (num_tracks, )
            curr_t: current frame's timestamps, torch.Tensor withwith shape (num_boxes,)

        Returns:
            classified edges: torch.Tensor with shape (num_steps, num_tracks, num_boxes),
                             containing at entry (step, i, j) the unnormalized probability that track i and
                             detection j are a match, according to the classifier at the given neural message passing step
        """

        # Get initial edge embeddings
        dist_reid =  cosine_distance(track_app, current_app).to(device)
        pos_edge_feats = self.compute_edge_feats(track_coords, current_coords, track_t, curr_t).to(device)

        edge_feats = torch.cat((pos_edge_feats, dist_reid.unsqueeze(-1)), dim=-1)
        edge_embeds = self.edge_in_mlp(edge_feats)
        initial_edge_embeds = edge_embeds.clone()

        # Get initial node embeddings, reduce dimensionality from 512 to node_dim
        track_embeds = F.relu(self.cnn_linear(track_app))
        curr_embeds =F.relu(self.cnn_linear(current_app))

        classified_edges = []
        for _ in range(self.num_steps):
            edge_embeds = torch.cat((edge_embeds, initial_edge_embeds), dim=-1)
            edge_embeds, track_embeds, curr_embeds = self.graph_net(edge_embeds=edge_embeds,
                                                                    nodes_a_embeds=track_embeds,
                                                                    nodes_b_embeds=curr_embeds)

            classified_edges.append(self.classifier(edge_embeds))

        return torch.stack(classified_edges).squeeze(-1)

## Putting everything together

Finally, we incorporate our ``AssignmentSimilarityNet`` into our tracker. We can keep everything as in ``LongTermReIDHungarianTracker`` except for the distance computation, which is now directly obtained via a forward pass through AssignmentSimilarityNet.

In [15]:
_UNMATCHED_COST=255
class MPNTracker(LongTermReIDHungarianTracker):
    def __init__(self, assign_net, *args, **kwargs):
        self.assign_net = assign_net
        super().__init__(*args, **kwargs)

    def data_association(self, boxes, scores, pred_features):
        if self.tracks:
            track_boxes = torch.stack([t.box for t in self.tracks], axis=0).cuda()
            track_features = torch.stack([t.get_feature() for t in self.tracks], axis=0).cuda()

            # Hacky way to recover the timestamps of boxes and tracks
            curr_t = self.im_index * torch.ones((pred_features.shape[0],)).cuda()
            track_t = torch.as_tensor([self.im_index - t.inactive - 1 for t in self.tracks]).cuda()

            ########################
            #### TODO starts
            ########################

            # Do a forward pass through self.assign_net to obtain our costs.
            # Note: self.assign_net will return unnormalized probabilities.
            # Make sure to apply the sigmoid function to them!

            pred_sim = self.assign_net(track_features.to(device), pred_features.to(device), track_boxes, boxes.to(device), track_t, curr_t)
            pred_sim = torch.sigmoid(pred_sim).detach().cpu()
            ########################
            #### TODO ends
            ########################

            pred_sim = pred_sim[-1]  # Use predictions at last message passing step
            distance = (1- pred_sim)

            # Do not allow mataches when sim < 0.5, to avoid low-confident associations
            distance = np.where(pred_sim < 0.5, _UNMATCHED_COST, distance)

            # Perform Hungarian matching.
            row_idx, col_idx = linear_assignment(distance)
            self.update_tracks(row_idx, col_idx, distance, boxes, scores, pred_features)


        else:
            # No tracks exist.
            self.add(boxes, scores, pred_features)

## Training and evaluating our model

We provide all boilerplate code for training our neural message passing based
tracker, as well as evaluating.

Under the hood, we are sampling frames randomly from our training sequences, and then sampling boxes from past frames as past_tracks to generate our
training
data. Check out `LongTrackTrainingDataset` for details.

We train the model with a weighted cross-entropy loss
to account for the class imbalance. Check out `train_one_epoch` if you're
interested.

Since the data is unbalanced, we add a weight $w$ for the positive samples in the binary cross entropy loss.
$$ \sum_{{i,j}} w \cdot y_{i,j} \log\left(\hat{y_{i,j}} \right) + \left(1-y_{i,j}\right)\log\left(1- \hat{y_{i,j}}\right), $$
where $$y_{i,j} \in\{ 0,1\}$$ is one, if the edge is active in the training data and zero, otherwise.
The variable $$\hat{y_{i,j}} \in[0,1]$$ denotes the learned output probability of the edge being active, which is later used for classification.


No need to write any code from your side here!


In [16]:
from gnn.dataset import LongTrackTrainingDataset
from torch.utils.data import DataLoader
from gnn.trainer import train_one_epoch

MAX_PATIENCE = 20
MAX_EPOCHS = 15
EVAL_FREQ = 1


# Define our model, and init
assign_net = AssignmentSimilarityNet(reid_network=None, # Not needed since we work with precomputed features
                                     node_dim=32,
                                     edge_dim=64,
                                     reid_dim=512,
                                     edges_in_dim=7,
                                     num_steps=10).cuda()

# We only keep two sequences for validation
dataset = LongTrackTrainingDataset(dataset='MOT16-train_wo_val2',
                                   db=train_db,
                                   root_dir= osp.join(root_dir, 'data/MOT16'),
                                   max_past_frames = MAX_PATIENCE,
                                   vis_threshold=0.25)

data_loader = DataLoader(dataset, batch_size=8, collate_fn = lambda x: x,
                         shuffle=True, num_workers=0, drop_last=True)
device = torch.device('cuda')
optimizer = torch.optim.Adam(assign_net.parameters(), lr=0.001)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=5)

We only leave 2 sequences for validation in order to maximize
the amount of training data. For your convenience, here are the
 LongTermReIDTracker results on them. Your validation IDF1 scores should show an improvement over them.
```

          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 45.9% 65.1% 35.4% 52.2% 96.1%  62 12 37 13 390  8873 130  210 49.4% 0.090
MOT16-11 68.3% 75.3% 62.5% 80.2% 96.6%  75 44 24  7 266  1871 136   90 75.9% 0.083
OVERALL  54.3% 69.6% 44.5% 61.7% 96.3% 137 56 61 20 656 10744 266  300 58.4% 0.087
```



Let's start training!

Note that we have observed quite a lot of noise in validation scores among epochs and runs. That is due to the small size of our training and validation sets. Therefore, we perform early stopping to obtain the best-performing model on validation. In addition, changing the experiment seed and/ or relaunching the training might help if you suspect noise influencing your scores.

In [17]:
best_idf1 = 0.
for epoch in range(1, MAX_EPOCHS + 1):
    print(f"-------- EPOC H {epoch:2d} --------")
    train_one_epoch(model = assign_net, data_loader=data_loader, optimizer=optimizer, print_freq=100)
    scheduler.step()

    if epoch % EVAL_FREQ == 0:
        tracker =  MPNTracker(assign_net=assign_net.eval(), obj_detect=None, patience=MAX_PATIENCE)
        val_sequences = MOT16Sequences('MOT16-val2', osp.join(root_dir, 'data/MOT16'), vis_threshold=0.)
        res = run_tracker(val_sequences, db=train_db, tracker=tracker, output_dir=None)
        idf1 = res.loc['OVERALL']['idf1']
        if idf1 > best_idf1:
            best_idf1 = idf1
            torch.save(assign_net.state_dict(), osp.join(root_dir, 'output', 'best_ckpt.pth'))


-------- EPOC H  1 --------


100it [06:26,  4.04s/it]

Iter 100. Loss: 0.583. Accuracy: 0.958. Recall: 0.739. Precision: 0.601


200it [13:07,  4.93s/it]

Iter 200. Loss: 0.065. Accuracy: 0.993. Recall: 0.989. Precision: 0.918


300it [19:42,  4.04s/it]

Iter 300. Loss: 0.042. Accuracy: 0.997. Recall: 0.994. Precision: 0.958


400it [26:01,  4.11s/it]

Iter 400. Loss: 0.029. Accuracy: 0.998. Recall: 0.996. Precision: 0.981


476it [31:03,  3.91s/it]


Tracking: MOT16-02
Tracks found: 92
Runtime for MOT16-02: 116.2 s.
Tracking: MOT16-11
Tracks found: 76
Runtime for MOT16-11: 47.0 s.
Runtime for all sequences: 163.2 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 45.3% 64.2% 34.9% 52.3% 96.1%  62 11 38 13 393  8865 103  224 49.6% 0.095
MOT16-11 69.2% 76.2% 63.3% 80.2% 96.6%  75 44 24  7 270  1868  44   88 76.9% 0.083
OVERALL  54.3% 69.5% 44.5% 61.7% 96.3% 137 55 62 20 663 10733 147  312 58.8% 0.090
-------- EPOC H  2 --------


100it [02:33,  1.65s/it]

Iter 100. Loss: 0.015. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


200it [05:09,  1.35s/it]

Iter 200. Loss: 0.017. Accuracy: 0.999. Recall: 0.998. Precision: 0.988


300it [07:39,  1.67s/it]

Iter 300. Loss: 0.014. Accuracy: 0.999. Recall: 0.999. Precision: 0.988


400it [10:25,  1.35s/it]

Iter 400. Loss: 0.019. Accuracy: 0.998. Recall: 0.998. Precision: 0.985


476it [12:32,  1.58s/it]


Tracking: MOT16-02
Tracks found: 67
Runtime for MOT16-02: 79.2 s.
Tracking: MOT16-11
Tracks found: 71
Runtime for MOT16-11: 47.0 s.
Runtime for all sequences: 126.2 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 43.9% 62.3% 33.9% 52.3% 96.1%  62 11 38 13 394  8867 103  223 49.6% 0.095
MOT16-11 68.1% 75.1% 62.3% 80.2% 96.6%  75 44 24  7 268  1868  43   88 76.9% 0.083
OVERALL  53.0% 67.9% 43.5% 61.7% 96.3% 137 55 62 20 662 10735 146  311 58.8% 0.090
-------- EPOC H  3 --------


100it [02:53,  1.82s/it]

Iter 100. Loss: 0.013. Accuracy: 0.999. Recall: 0.998. Precision: 0.988


200it [05:56,  1.66s/it]

Iter 200. Loss: 0.018. Accuracy: 0.999. Recall: 0.998. Precision: 0.985


300it [08:57,  1.70s/it]

Iter 300. Loss: 0.014. Accuracy: 0.999. Recall: 0.998. Precision: 0.986


400it [11:46,  1.68s/it]

Iter 400. Loss: 0.283. Accuracy: 0.993. Recall: 0.927. Precision: 0.920


476it [13:56,  1.76s/it]


Tracking: MOT16-02
Tracks found: 189
Runtime for MOT16-02: 87.9 s.
Tracking: MOT16-11
Tracks found: 121
Runtime for MOT16-11: 47.4 s.
Runtime for all sequences: 135.3 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 47.7% 67.7% 36.9% 52.3% 96.1%  62 11 38 13 396  8861 118  211 49.5% 0.095
MOT16-11 74.6% 82.2% 68.3% 80.2% 96.6%  75 44 24  7 269  1869  47   88 76.8% 0.083
OVERALL  57.8% 74.0% 47.4% 61.7% 96.3% 137 55 62 20 665 10730 165  299 58.7% 0.089
-------- EPOC H  4 --------


100it [02:34,  1.14s/it]

Iter 100. Loss: 0.028. Accuracy: 0.999. Recall: 0.997. Precision: 0.989


200it [05:17,  1.70s/it]

Iter 200. Loss: 0.021. Accuracy: 0.999. Recall: 0.997. Precision: 0.990


300it [08:01,  1.81s/it]

Iter 300. Loss: 0.017. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


400it [10:42,  1.51s/it]

Iter 400. Loss: 0.019. Accuracy: 0.999. Recall: 0.997. Precision: 0.989


476it [12:44,  1.61s/it]


Tracking: MOT16-02
Tracks found: 93
Runtime for MOT16-02: 73.5 s.
Tracking: MOT16-11
Tracks found: 81
Runtime for MOT16-11: 44.4 s.
Runtime for all sequences: 117.9 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 46.5% 66.0% 35.9% 52.3% 96.1%  62 11 38 13 394  8863 100  216 49.6% 0.095
MOT16-11 68.4% 75.4% 62.6% 80.2% 96.6%  75 44 24  7 267  1869  46   89 76.9% 0.083
OVERALL  54.7% 70.1% 44.9% 61.7% 96.3% 137 55 62 20 661 10732 146  305 58.8% 0.090
-------- EPOC H  5 --------


100it [02:25,  1.63s/it]

Iter 100. Loss: 0.015. Accuracy: 0.999. Recall: 0.998. Precision: 0.987


200it [04:54,  1.49s/it]

Iter 200. Loss: 0.015. Accuracy: 0.999. Recall: 0.998. Precision: 0.987


300it [07:20,  1.47s/it]

Iter 300. Loss: 0.014. Accuracy: 0.999. Recall: 0.998. Precision: 0.988


400it [09:49,  1.45s/it]

Iter 400. Loss: 0.014. Accuracy: 0.999. Recall: 0.998. Precision: 0.988


476it [11:39,  1.47s/it]


Tracking: MOT16-02
Tracks found: 88
Runtime for MOT16-02: 70.3 s.
Tracking: MOT16-11
Tracks found: 81
Runtime for MOT16-11: 43.7 s.
Runtime for all sequences: 114.1 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 45.8% 65.0% 35.3% 52.3% 96.1%  62 11 38 13 393  8864 103  216 49.6% 0.095
MOT16-11 69.0% 76.1% 63.2% 80.2% 96.6%  75 44 24  7 268  1869  45   89 76.9% 0.083
OVERALL  54.5% 69.8% 44.7% 61.7% 96.3% 137 55 62 20 661 10733 148  305 58.8% 0.090
-------- EPOC H  6 --------


100it [02:28,  1.65s/it]

Iter 100. Loss: 0.011. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


200it [04:58,  1.50s/it]

Iter 200. Loss: 0.013. Accuracy: 0.999. Recall: 0.998. Precision: 0.990


300it [07:18,  1.22s/it]

Iter 300. Loss: 0.010. Accuracy: 0.999. Recall: 0.998. Precision: 0.990


400it [09:37,  1.39s/it]

Iter 400. Loss: 0.013. Accuracy: 0.999. Recall: 0.999. Precision: 0.988


476it [11:31,  1.45s/it]


Tracking: MOT16-02
Tracks found: 91
Runtime for MOT16-02: 70.3 s.
Tracking: MOT16-11
Tracks found: 81
Runtime for MOT16-11: 42.7 s.
Runtime for all sequences: 112.9 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 46.7% 66.2% 36.0% 52.3% 96.1%  62 11 38 13 393  8862  99  219 49.7% 0.095
MOT16-11 69.2% 76.2% 63.3% 80.2% 96.6%  75 44 24  7 268  1869  44   89 76.9% 0.083
OVERALL  55.1% 70.6% 45.2% 61.7% 96.3% 137 55 62 20 661 10731 143  308 58.8% 0.090
-------- EPOC H  7 --------


100it [02:24,  1.75s/it]

Iter 100. Loss: 0.013. Accuracy: 0.999. Recall: 0.998. Precision: 0.990


200it [04:52,  1.53s/it]

Iter 200. Loss: 0.009. Accuracy: 0.999. Recall: 0.999. Precision: 0.991


300it [07:17,  1.35s/it]

Iter 300. Loss: 0.009. Accuracy: 0.999. Recall: 0.999. Precision: 0.991


400it [09:40,  1.53s/it]

Iter 400. Loss: 0.013. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


476it [11:27,  1.44s/it]


Tracking: MOT16-02
Tracks found: 94
Runtime for MOT16-02: 70.0 s.
Tracking: MOT16-11
Tracks found: 82
Runtime for MOT16-11: 43.1 s.
Runtime for all sequences: 113.0 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 47.2% 67.0% 36.5% 52.3% 96.1%  62 11 38 13 393  8863 100  219 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1869  43   89 76.9% 0.083
OVERALL  55.6% 71.2% 45.6% 61.7% 96.3% 137 55 62 20 660 10732 143  308 58.8% 0.090
-------- EPOC H  8 --------


100it [02:32,  1.47s/it]

Iter 100. Loss: 0.009. Accuracy: 0.999. Recall: 1.000. Precision: 0.992


200it [04:57,  1.53s/it]

Iter 200. Loss: 0.009. Accuracy: 0.999. Recall: 0.999. Precision: 0.992


300it [07:13,  1.41s/it]

Iter 300. Loss: 0.013. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


400it [09:37,  1.40s/it]

Iter 400. Loss: 0.013. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


476it [11:26,  1.44s/it]


Tracking: MOT16-02
Tracks found: 88
Runtime for MOT16-02: 70.0 s.
Tracking: MOT16-11
Tracks found: 81
Runtime for MOT16-11: 42.5 s.
Runtime for all sequences: 112.5 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 46.2% 65.5% 35.6% 52.3% 96.1%  62 11 38 13 393  8864 102  219 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 268  1869  43   89 76.9% 0.083
OVERALL  54.9% 70.3% 45.0% 61.7% 96.3% 137 55 62 20 661 10733 145  308 58.8% 0.090
-------- EPOC H  9 --------


100it [02:24,  1.61s/it]

Iter 100. Loss: 0.011. Accuracy: 0.999. Recall: 0.999. Precision: 0.991


200it [04:51,  1.29s/it]

Iter 200. Loss: 0.016. Accuracy: 0.999. Recall: 0.998. Precision: 0.987


300it [07:20,  1.25s/it]

Iter 300. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.987


400it [09:37,  1.33s/it]

Iter 400. Loss: 0.013. Accuracy: 0.999. Recall: 0.998. Precision: 0.990


476it [11:28,  1.45s/it]


Tracking: MOT16-02
Tracks found: 88
Runtime for MOT16-02: 70.0 s.
Tracking: MOT16-11
Tracks found: 82
Runtime for MOT16-11: 43.1 s.
Runtime for all sequences: 113.1 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 46.7% 66.3% 36.1% 52.3% 96.1%  62 11 38 13 393  8864  99  220 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1869  41   89 76.9% 0.083
OVERALL  55.3% 70.8% 45.3% 61.7% 96.3% 137 55 62 20 660 10733 140  309 58.8% 0.090
-------- EPOC H 10 --------


100it [02:27,  1.41s/it]

Iter 100. Loss: 0.009. Accuracy: 0.999. Recall: 1.000. Precision: 0.991


200it [04:49,  1.38s/it]

Iter 200. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.990


300it [07:13,  1.78s/it]

Iter 300. Loss: 0.009. Accuracy: 0.999. Recall: 0.999. Precision: 0.992


400it [09:32,  1.28s/it]

Iter 400. Loss: 0.011. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


476it [11:24,  1.44s/it]


Tracking: MOT16-02
Tracks found: 89
Runtime for MOT16-02: 69.9 s.
Tracking: MOT16-11
Tracks found: 82
Runtime for MOT16-11: 42.7 s.
Runtime for all sequences: 112.6 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 47.8% 67.8% 36.9% 52.3% 96.1%  62 11 38 13 393  8864  99  220 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1869  41   89 76.9% 0.083
OVERALL  55.9% 71.6% 45.9% 61.7% 96.3% 137 55 62 20 660 10733 140  309 58.8% 0.090
-------- EPOC H 11 --------


100it [02:27,  1.28s/it]

Iter 100. Loss: 0.012. Accuracy: 0.999. Recall: 0.998. Precision: 0.987


200it [04:42,  1.39s/it]

Iter 200. Loss: 0.009. Accuracy: 0.999. Recall: 0.999. Precision: 0.989


300it [07:08,  1.46s/it]

Iter 300. Loss: 0.008. Accuracy: 0.999. Recall: 0.999. Precision: 0.992


400it [09:31,  1.34s/it]

Iter 400. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.989


476it [11:24,  1.44s/it]


Tracking: MOT16-02
Tracks found: 89
Runtime for MOT16-02: 70.5 s.
Tracking: MOT16-11
Tracks found: 82
Runtime for MOT16-11: 43.0 s.
Runtime for all sequences: 113.5 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 47.8% 67.8% 36.9% 52.3% 96.1%  62 11 38 13 393  8864  99  220 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1869  41   89 76.9% 0.083
OVERALL  55.9% 71.6% 45.9% 61.7% 96.3% 137 55 62 20 660 10733 140  309 58.8% 0.090
-------- EPOC H 12 --------


100it [02:20,  1.39s/it]

Iter 100. Loss: 0.012. Accuracy: 0.999. Recall: 0.999. Precision: 0.990


200it [04:44,  1.41s/it]

Iter 200. Loss: 0.011. Accuracy: 0.999. Recall: 0.999. Precision: 0.992


300it [07:08,  1.42s/it]

Iter 300. Loss: 0.007. Accuracy: 0.999. Recall: 1.000. Precision: 0.990


400it [09:32,  1.53s/it]

Iter 400. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.990


476it [11:25,  1.44s/it]


Tracking: MOT16-02
Tracks found: 89
Runtime for MOT16-02: 70.1 s.
Tracking: MOT16-11
Tracks found: 80
Runtime for MOT16-11: 43.3 s.
Runtime for all sequences: 113.4 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 46.8% 66.5% 36.2% 52.3% 96.1%  62 11 38 13 393  8865 101  220 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1870  41   89 76.9% 0.083
OVERALL  55.3% 70.9% 45.4% 61.7% 96.3% 137 55 62 20 660 10735 142  309 58.8% 0.090
-------- EPOC H 13 --------


100it [02:21,  1.22s/it]

Iter 100. Loss: 0.012. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


200it [04:42,  1.46s/it]

Iter 200. Loss: 0.012. Accuracy: 0.999. Recall: 0.997. Precision: 0.987


300it [07:09,  1.51s/it]

Iter 300. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.989


400it [09:34,  1.38s/it]

Iter 400. Loss: 0.008. Accuracy: 0.999. Recall: 0.999. Precision: 0.991


476it [11:25,  1.44s/it]


Tracking: MOT16-02
Tracks found: 89
Runtime for MOT16-02: 70.6 s.
Tracking: MOT16-11
Tracks found: 82
Runtime for MOT16-11: 42.3 s.
Runtime for all sequences: 112.9 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 47.8% 67.8% 36.9% 52.3% 96.1%  62 11 38 13 393  8864  99  220 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1869  41   89 76.9% 0.083
OVERALL  55.9% 71.6% 45.9% 61.7% 96.3% 137 55 62 20 660 10733 140  309 58.8% 0.090
-------- EPOC H 14 --------


100it [02:26,  1.62s/it]

Iter 100. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.988


200it [04:52,  1.39s/it]

Iter 200. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.989


300it [07:13,  1.49s/it]

Iter 300. Loss: 0.013. Accuracy: 0.999. Recall: 0.997. Precision: 0.987


400it [09:33,  1.27s/it]

Iter 400. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.989


476it [11:23,  1.44s/it]


Tracking: MOT16-02
Tracks found: 89
Runtime for MOT16-02: 69.7 s.
Tracking: MOT16-11
Tracks found: 82
Runtime for MOT16-11: 43.3 s.
Runtime for all sequences: 112.9 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 47.2% 66.9% 36.4% 52.3% 96.1%  62 11 38 13 393  8864  99  220 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1869  41   89 76.9% 0.083
OVERALL  55.5% 71.1% 45.5% 61.7% 96.3% 137 55 62 20 660 10733 140  309 58.8% 0.090
-------- EPOC H 15 --------


100it [02:29,  1.83s/it]

Iter 100. Loss: 0.011. Accuracy: 0.999. Recall: 0.998. Precision: 0.989


200it [04:47,  1.37s/it]

Iter 200. Loss: 0.010. Accuracy: 0.999. Recall: 0.999. Precision: 0.989


300it [07:07,  1.59s/it]

Iter 300. Loss: 0.009. Accuracy: 0.999. Recall: 0.999. Precision: 0.990


400it [09:29,  1.56s/it]

Iter 400. Loss: 0.012. Accuracy: 0.999. Recall: 0.999. Precision: 0.990


476it [11:23,  1.44s/it]


Tracking: MOT16-02
Tracks found: 89
Runtime for MOT16-02: 70.2 s.
Tracking: MOT16-11
Tracks found: 82
Runtime for MOT16-11: 42.6 s.
Runtime for all sequences: 112.8 s.


C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)
C:\Users\yusuf\anaconda3\lib\site-packages\motmetrics\mot.py:344: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  r = r.append(copy)


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 47.2% 66.9% 36.4% 52.3% 96.1%  62 11 38 13 393  8864  99  220 49.6% 0.095
MOT16-11 69.4% 76.5% 63.5% 80.2% 96.6%  75 44 24  7 267  1869  41   89 76.9% 0.083
OVERALL  55.5% 71.1% 45.5% 61.7% 96.3% 137 55 62 20 660 10733 140  309 58.8% 0.090


          IDF1   IDP   IDR  Rcll  Prcn  GT MT PT ML  FP    FN IDs   FM  MOTA  MOTP
MOT16-02 44.2% 62.8% 34.1% 52.3% 96.1%  62 11 38 13 391  8865 124  217 49.5% 0.095
MOT16-11 66.7% 73.5% 61.0% 80.2% 96.6%  75 44 24  7 267  1871  35   90 77.0% 0.083
OVERALL  52.7% 67.4% 43.2% 61.7% 96.3% 137 55 62 20 658 10736 159  307 58.8% 0.090

# Exercise submission

After executing the following cell, the `Colab Notebooks/cv3dst_exercise/output` directory in your Google Drive should contain multiple `MOT16-XY.txt` files.

For the final submission, you must process the test sequences and upload the zipped prediction files to our server.

Note that this time, you only have to evaluate three sequences!

In [19]:
best_ckpt = torch.load(osp.join(root_dir, 'output', 'best_ckpt.pth'))
assign_net.load_state_dict(best_ckpt)

tracker =  MPNTracker(assign_net=assign_net.eval(), obj_detect=None, patience=MAX_PATIENCE)
test_db = torch.load(osp.join(gnn_root_dir, 'data/preprocessed_data_test_2.pth'))
val_sequences = MOT16Sequences('MOT16-test', osp.join(root_dir, 'data/MOT16'), vis_threshold=0.)
run_tracker(val_sequences, db=test_db, tracker=tracker, output_dir=osp.join(root_dir, 'output'))

Tracking: MOT16-01
No GT evaluation data available.
Tracks found: 128
Runtime for MOT16-01: 38.6 s.
Writing predictions to: cv3dst_exercise/output\MOT16-01.txt
Tracking: MOT16-08
No GT evaluation data available.
Tracks found: 264
Runtime for MOT16-08: 67.8 s.
Writing predictions to: cv3dst_exercise/output\MOT16-08.txt
Tracking: MOT16-12
No GT evaluation data available.
Tracks found: 188
Runtime for MOT16-12: 38.0 s.
Writing predictions to: cv3dst_exercise/output\MOT16-12.txt
Runtime for all sequences: 144.3 s.
